edgedb/abstract.py

```diff
@@ -48,6 +55,48 @@ class ExecuteContext(typing.NamedTuple):
     state: typing.Optional[options.State]
 
 
+class DescribeContext(typing.NamedTuple):
+    query: str
+    state: typing.Optional[options.State]
+
+
+class AbstractCodec:
+    @property
+    def codec_type(self) -> typing.Optional[enums.CodecType]:
+        return None
+
+    @property
+    def type_id(self) -> uuid.UUID:
+        raise NotImplementedError
+
+    @property
+    def type_name(self) -> typing.Optional[str]:
+        return None
+
+    @property
+    def base_type_name(self) -> typing.Optional[str]:
+        return None
+
+    @property
+    def sub_types(self) -> typing.Tuple[SubType]:
+        raise NotImplementedError
+
+
+class SubType:
+    codec: AbstractCodec
+    name: typing.Optional[str]
+    cardinality: typing.Optional[enums.Cardinality]
+    is_implicit: bool
+    is_link_property: bool
+    is_link: bool
```

Comment:
This is actually defined in `edgedb/introspect.py` already, reusing somehow is preferred.